File: app/llm_trace_test.py
"""Трасса вызовов моделей: запись, ротация, чтение обоих источников, hook в llm_client."""
from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

import llm_client
import llm_trace


def test_record_and_recent_merge_python_and_electron_sources(tmp_path, monkeypatch):
    monkeypatch.setattr(llm_trace, "DATA_ROOT", tmp_path)
    entry = llm_trace.record(request_id="r1", role="quality_judge", provider="claude", model="opus", effort="high",
                             contract_version="agent-contract/1.0", duration_ms=1234.4, prompt_chars=40, output_chars=10,
                             dropped=[{"field": "temperature", "reason": "x"}], structured_output=True)
    assert entry["source"] == "python" and entry["ok"] is True and entry["durationMs"] == 1234
    assert entry["dropped"] == ["temperature"] and entry["structuredOutput"] is True
    assert llm_trace.write(entry)
    electron = tmp_path / "traces" / llm_trace.ELECTRON_FILE
    electron.write_text(json.dumps({"ts": "2099-01-01T00:00:00Z", "requestId": "e1", "provider": "codex"}) + "\n"
                        + "not json\n", encoding="utf-8")
    recent = llm_trace.recent(10)
    assert [item["requestId"] for item in recent] == ["e1", "r1"], "новые первыми, битые строки пропускаются"
    assert recent[0]["source"] == "electron" and recent[1]["source"] == "python"
    assert llm_trace.recent(10, source="python")[0]["requestId"] == "r1"
    assert llm_trace.recent(1) and len(llm_trace.recent(1)) == 1


def test_write_rotates_at_size_cap_and_never_raises(tmp_path, monkeypatch):
    monkeypatch.setattr(llm_trace, "DATA_ROOT", tmp_path)
    monkeypatch.setattr(llm_trace, "MAX_BYTES", 120)
    for index in range(6):
        assert llm_trace.write({"ts": f"2026-09-09T00:00:0{index}Z", "requestId": f"r{index}"})
    directory = tmp_path / "traces"
    assert (directory / (llm_trace.PYTHON_FILE + ".1")).exists()
    assert llm_trace.recent(50)[0]["requestId"] == "r5"
    monkeypatch.setattr(llm_trace, "DATA_ROOT", tmp_path / "file-instead-of-dir")
    (tmp_path / "file-instead-of-dir").write_text("x", encoding="utf-8")
    assert llm_trace.write({"ts": "x"}) is False


def test_message_chars_counts_text_parts_only():
    messages = [{"role": "user", "content": "abcd"}, {"role": "user", "content": [
        {"type": "text", "text": "xyz"}, {"type": "image_url", "image_url": {"url": "data:image/png;base64,AAAA"}}]}]
    assert llm_trace.message_chars(messages, "SYS") == 3 + 4 + 3


def test_chat_envelope_writes_a_trace_for_cli_calls_and_failures(monkeypatch, tmp_path):
    monkeypatch.setenv("DESIGNDNA_CLAUDE", str(tmp_path / "claude.exe"))
    (tmp_path / "claude.exe").write_text("")
    monkeypatch.delenv("OPENAI_API_KEY", raising=False)
    written = []
    monkeypatch.setattr(llm_trace, "write", lambda entry, data_root=None: written.append(entry) or True)
    monkeypatch.setattr(llm_client.cli_llm, "chat", lambda provider, messages, **kw: "{\"ok\": true}")
    request = llm_client.ChatRequest(messages=[{"role": "user", "content": "hello"}], provider="claude",
                                     system="SYS", reasoning_effort="high",
                                     response_format="json_schema", response_json_schema={"name": "v", "schema": {"type": "object"}})
    llm_client.chat_envelope(request, role="quality_judge")
    entry = written[-1]
    assert entry["provider"] == "claude" and entry["profile"] == "quality_judge" and entry["effort"] == "high"
    assert entry["ok"] is True and entry["outputChars"] == len('{"ok": true}') and entry["promptChars"] == len("SYS") + len("hello")
    assert entry["structuredOutput"] is True and entry["contractVersion"].startswith("agent-contract/")
    assert "hello" not in json.dumps(entry) and "SYS" not in json.dumps(entry)

    def boom(provider, messages, **kw):
        raise RuntimeError("Claude Code: not logged in")

    monkeypatch.setattr(llm_client.cli_llm, "chat", boom)
    try:
        llm_client.chat_envelope(llm_client.ChatRequest(messages=[{"role": "user", "content": "x"}], provider="claude"), role="generator")
    except RuntimeError:
        pass
    failed = written[-1]
    assert failed["ok"] is False and "not logged in" in failed["error"] and failed["profile"] == "generator"

File: desktop/tests/llm-trace.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { existsSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

import { LlmTrace, MAX_TRACE_BYTES, TRACE_FILE_NAME, messageChars, traceRecord } from "../services/llm-trace.mjs";

function scratch(t) {
  const dir = mkdtempSync(path.join(tmpdir(), "ddna-trace-"));
  t.after(() => { try { rmSync(dir, { recursive: true, force: true }); } catch { /* best effort */ } });
  return dir;
}

test("a trace record keeps metadata only and never the prompt or the answer", () => {
  const envelope = {
    id: "req-1", provider: "claude", model: "opus", system: "SECRET SYSTEM PROMPT",
    reasoning: { effort: "high" },
    messages: [{ role: "user", content: "SECRET USER TEXT" }, { role: "user", content: [{ type: "text", text: "12345" }, { type: "image_url", image_url: { url: "data:image/png;base64,AAAA" } }] }],
  };
  const result = { content: '{"approved":true}', provider: "claude",
    transport: { provider: "claude", model: "opus", contractVersion: "agent-contract/1.0", structuredOutput: true,
      dropped: [{ field: "temperature", reason: "x" }], instructionSources: ["a", "b"], threadId: "t1", turnId: "u1" } };
  const record = traceRecord({ envelope, profile: "quality_judge", result, durationMs: 1234.6, queuedMs: 12 });
  assert.equal(record.source, "electron");
  assert.equal(record.requestId, "req-1");
  assert.equal(record.profile, "quality_judge");
  assert.equal(record.provider, "claude");
  assert.equal(record.model, "opus");
  assert.equal(record.effort, "high");
  assert.equal(record.contractVersion, "agent-contract/1.0");
  assert.equal(record.structuredOutput, true);
  assert.deepEqual(record.dropped, ["temperature"]);
  assert.equal(record.durationMs, 1235);
  assert.equal(record.queuedMs, 12);
  assert.equal(record.ok, true);
  assert.equal(record.error, null);
  assert.equal(record.promptChars, "SECRET SYSTEM PROMPT".length + "SECRET USER TEXT".length + 5);
  assert.equal(record.outputChars, 17);
  assert.equal(record.instructionSources, 2);
  assert.equal(record.threadId, "t1");
  const serialized = JSON.stringify(record);
  assert.doesNotMatch(serialized, /SECRET/);
  assert.doesNotMatch(serialized, /approved/);
  assert.doesNotMatch(serialized, /base64/);
  assert.equal(messageChars([{ role: "user", content: null }]), 0);
});

test("errors and cancellations are recorded with the failure, not the payload", () => {
  const record = traceRecord({ envelope: { id: "req-2", provider: "codex", messages: [] }, profile: "generator",
    error: new Error("Codex request cancelled"), durationMs: 5, cancelled: true });
  assert.equal(record.ok, false);
  assert.equal(record.cancelled, true);
  assert.equal(record.error, "Codex request cancelled");
  assert.equal(record.provider, "codex");
  assert.equal(record.outputChars, 0);
});

test("writes JSON lines into <data>/traces and rotates the file at the size cap", (t) => {
  const dir = path.join(scratch(t), "data", "traces");
  const trace = new LlmTrace({ directory: dir, maxBytes: 200 });
  assert.equal(trace.write({ ts: "2026-09-09T00:00:00Z", ok: true }), true);
  const file = path.join(dir, TRACE_FILE_NAME);
  assert.ok(existsSync(file));
  assert.equal(readFileSync(file, "utf8").trim().split("\n").length, 1);
  writeFileSync(file, "x".repeat(250));
  assert.equal(trace.write({ ts: "2026-09-09T00:00:01Z", ok: false }), true);
  assert.ok(existsSync(`${file}.1`), "the oversized file is rotated aside");
  assert.equal(readFileSync(file, "utf8").trim().split("\n").length, 1);
  assert.ok(MAX_TRACE_BYTES > 1024 * 1024);
});

test("a failing writer never throws into the provider call", (t) => {
  const errors = [];
  const trace = new LlmTrace({ directory: path.join(scratch(t), "traces"),
    appendFile: () => { throw new Error("disk full"); }, onError: (error) => errors.push(error.message) });
  assert.equal(trace.write({ ok: true }), false);
  assert.deepEqual(errors, ["disk full"]);
});
